### DISCUSSION TITLE
Air gap installation and GC

### DISCUSSION THREAD

**Discussion by sushiMix** [k3s-io/k3s#8779] [General]
Hello, I know there are several tickets about it but didn't find a clear view.

Accordingly to this ticket https://github.com/k3s-io/k3s/issues/3193 it seems working from k3s 1.24
but https://github.com/k3s-io/k3s/issues/7569 there is no possible way to protect images.
Accordingly to test It seems still not working.

But I have seen tickets in k8s/containerd :
https://github.com/containerd/containerd/issues/6352 and https://github.com/containerd/containerd/issues/6930

and did a small test:

 Use ctr to load an image and add the pinning label
` sudo ./ctr images label "my image" io.cri-containerd.pinned=pinned`

Fill the disk using `xfs_mkfile` like the first ticket

The image with the pin tag is no more freed when GC is run (like the pause image :)).
I'm using k3s version v1.26.9+k3s1.

Isn't it a solution to protect some images from GC ?

**Comment by brandond** [comment]
yes, the leases that we added a while back only protected the layer content, they did not protect the tags as hoped. We should rework our protections to use the CRI pinned label.

**Reply by sushiMix** [reply]
There should be also a way to cleanup these flag after an upgrade to remove old images.

**Reply by brandond** [reply]
With the leases, we manage it by only holding leases on whatever was most recently loaded from the tarballs. I'd need to look at how this works on the backend to see if we can do this with the pinned annotation. If something like `io.cri-containerd.pinned=k3s` would work that would make it easy.
